Validate the incoming GitHub webhook request

kshitij10496 opened the pull request:
Using a custom HMAC Authenticator and a new `GITHUB_SECRET` env variable,
we can now validate that the webhook request is indeed initiated by GitHub.

**NOTE:** The `GITHUB_SECRET` is now required for running the server and must be added to the`.env` file.

Fixes #18 

Suggestions and/or feedback on improving the implementation are welcome! 😄

kshitij10496 commented:
@BRMatt Can you take a look at this PR? 😅

BRMatt commented:
Hey @kshitij10496, sorry I was on vacation for a few weeks and only just got back. I'll try to review this later today, but it seems like this PR is causing the tests in `routes_test.go` to fail? We should really have circleci enabled on this repo, I'll try to make that happen.

kshitij10496 commented:
Hi @BRMatt!
I too was AFK the past couple of weeks - that evens us out. 😅 
Nonetheless, I will fix the failing tests right away.

> We should really have circleci enabled on this repo, I'll try to make that happen.

Most definitely! 
I noticed that we already have a CircleCI configuration file in the repo.
Would it not be a matter of simply enable GitHub checks on the repo now? 
Or is there something that I'm overlooking here? 🤔

kshitij10496 commented:
Fixed the tests! 😃 

I've implemented a `WebhookValidator` interface which can be used to validate the incoming requests. In particular, the `GitHubWebhookValidator` type helps us secure the incoming webhook requests from GitHub.

**Note:** With the additional LoC, the code coverage has decreased from 19.8% to 17.8%.

<img width="1280" alt="go test -cover" src="https://user-images.githubusercontent.com/12427142/50345933-bd229f80-0556-11e9-88d9-fddc26cac9d7.png">

BRMatt commented on validator.go:
```diff
@@ -0,0 +1,62 @@
+// GitHubWebhookValidator implements the WebhookValidator interface.
+// The implementation uses HMAC encryption-decryption using the "secret" key.
+//
+// Refer: https://developer.github.com/webhooks/securing/#securing-your-webhooks
```
Thanks for referencing the docs here - very useful!

BRMatt commented on validator.go:
```diff
@@ -0,0 +1,62 @@
+	}
+	defer r.Body.Close()
+
+	b, err := ioutil.ReadAll(r.Body)
```
Hmm, I have a feeling that reading the entire body and `defer`ing its closure will cause the bot to break as the HTTP handler will not be able to read the body after its content has been read, and the reader has been closed. You will probably need to [create a new reader from the bytes that you have read, and assign it to `r.Body`](https://medium.com/@xoen/golang-read-from-an-io-readwriter-without-loosing-its-content-2c6911805361).

I think the fake validator is preventing the tests from highlighting this problem. 😔Maybe we should use the real validator in the tests? Or at least, have a test that verifies we can still use the request body after verifying the request.

kshitij10496 replied:
> Hmm, I have a feeling that reading the entire body and `defer`ing its closure will cause the bot to break as the HTTP handler will not be able to read the body after its content has been read, and the reader has been closed.

You are absolutely correct! 
I have most definitely overlooked the fact that we can only read from an `io.ReadCloser` just once. Thanks a lot for catching this in the review. 👍 

I agree with your suggestion of reassigning the request body after performing the validation. 
I will use the following approach as described in the associated Medium article:
1. Instantiate a new reader: `bytes.NewReader()` to obtain an `io.Reader` from the `byte` slice.
2. Create a new `io.NopCloser` for the above reader. 
3. Assign the newly created `io.ReadCloser` to the request body.

The only thing I have changed is the use of `bytes.NewReader()` in favour of `bytes.NewBuffer()`. I think we should use `bytes.NewReader()` because creating a `bytes.Reader` value allows us for the bare minimum functionality required for the task while a `byes.Buffer` value provides us with additional, though unnecessary, `Write` methods to manipulate the buffer itself. However, I haven't worked out the benchmarks (speed and memory) for either value. Thus, I would like to know your preference and implement accordingly. 😄

BRMatt replied:
`bytes.NewReader()` seems like a good choice, you're right that we don't need the extra write methods. I don't think it's worth benchmarking the difference as this part of the code is unlikely to be the bottleneck - the api calls are much more costly in terms of response time 😂

kshitij10496 commented:
Wish you a very happy new year too, Matt! 🎉 😄 

> presumably to prevent folks mining bitcoin through open source projects!

 😆 I hope that is the case.

> There's one small issue that we need to resolve before merging this in, but I can do it for you if you're tired of working on this.

No worries! Let's see this through. 👍

kshitij10496 commented on routes.go:
```diff
@@ -12,9 +12,10 @@ import (
 )
 
-func NewServer(notifier Notifier) http.Handler {
+func NewServer(notifier Notifier, validator WebhookValidator) http.Handler {
```
This looks like a bug in the bot, doesn't it? 🤔

fatih replied:
Yes this is a bug :) I'll fix it tonight, thanks !

kshitij10496 replied:
Thanks, @fatih! 😄

kshitij10496 commented:
@BRMatt, Let me know if there is anything more needs to be improved in this PR prior to merging.
I had almost forgotten about this! 😅

BRMatt commented:
@kshitij10496 Sorry about the delay, it's sometimes hard to set aside time. Thanks again for doing this work, I appreciate it! 🙇 🎉

kshitij10496 commented:
@BRMatt Thanks a lot for the merge! 🎉 
I am glad I could contribute. 😄
